rust/src/commands/transaction.rs

```diff
@@ -0,0 +1,171 @@
+//! Wrapper Execution plan to handle distributed operations
+use super::*;
+use crate::action::Action;
+// use crate::commands::create::CreateCommand;
+use async_trait::async_trait;
+use core::any::Any;
+use datafusion::{
+    arrow::{
+        array::{Array, StringArray},
+        datatypes::{
+            DataType, Field as ArrowField, Schema as ArrowSchema, SchemaRef as ArrowSchemaRef,
+        },
+        record_batch::RecordBatch,
+    },
+    error::Result as DataFusionResult,
+    physical_plan::{
+        coalesce_partitions::CoalescePartitionsExec, common::compute_record_batch_statistics,
+        empty::EmptyExec, Distribution, ExecutionPlan, Partitioning, SendableRecordBatchStream,
+        Statistics,
+    },
+};
+use lazy_static::lazy_static;
+use std::sync::Arc;
+
+lazy_static! {
+    /// Schema expected form plans wrapped by transaction
+    pub static ref OPERATION_SCHEMA: ArrowSchema =
+        ArrowSchema::new(vec![ArrowField::new("serialized", DataType::Utf8, true,)]);
+}
+
+pub(crate) fn serialize_actions(actions: Vec<Action>) -> DataFusionResult<RecordBatch> {
+    let serialized = StringArray::from(
+        actions
+            .iter()
+            .map(serde_json::to_string)
+            .collect::<Result<Vec<_>, _>>()
+            .map_err(to_datafusion_err)?,
+    );
+    Ok(RecordBatch::try_new(
+        Arc::new(OPERATION_SCHEMA.clone()),
+        vec![Arc::new(serialized)],
+    )?)
+}
+
+/// Write command
+#[derive(Debug)]
+pub struct DeltaTransactionPlan {
+    table_uri: String,
+    input: Arc<dyn ExecutionPlan>,
```

Comment:
One problem I noticed with `DeltaTransactionPlan` is transaction actions need to be prepared for a specific version of a table so we can detect and resolve transaction conflicts. However, within its execute method, we always load the latest version of the table, which might be newer than the version we used to generate the actions.

For example, consider an overwrite table write actions generated for version 1, then another writer wrote a bunch of new files into version 2 before we commit the overwrite write. When we try to commit the overwrite transaction as version 3, we will miss the new files that got added in version 2 resulting in incorrect table state.

One way to solve this problem is to add a table version attribute to `DeltaTransactionPlan`. Alternative, it might be better to get rid of the `DeltaTransactionPlan` entirely and let individual command execution plans manage the transaction struct internally. A table transaction is just a simple json file write, so there is not much value in managing this write as an execution plan. Execution plan is especially useful for operations needs to be executed in a distributed fashion, for example, writing out partitioned parquet files for each add action.

Reply:
Writing this I was also a bit conflicted about how to best handle wrapping things in a transaction. Making sure we commit against a fixed version seems reasonable. For my understanding though, in the example above we would have to update the log to the latest version before retrying, right? 

The reason this found its way into an execution plan is, twofold. For one i tried to mimic what spark does as best I could. There they seem to have something that conceptually looks a bit like a contextmanager in python.

```scala
override def run(sparkSession: SparkSession): Seq[Row] = {
    deltaLog.withNewTransaction { txn =>

      val actions = write(txn, sparkSession)
      val operation = DeltaOperations.Write(mode, Option(partitionColumns),
        options.replaceWhere, options.userMetadata)
      txn.commit(actions, operation)
    }
    Seq.empty
  }
```
While I may have misunderstood whats going on there, I tried to come up with an implementation that could be used in  a similar fashion, i.e. collect all actions from child plans and commit them based on the operation. Of course we can achieve this without all the overhead of an execution plan. However if we wanted to support the Delta Specific SQL syntax from spark in the future based on datafusion / ballista, I thought wrapping things in an execution plan would be the mst straight forward way to go. But again I may have missed something :)

Reply:
> For my understanding though, in the example above we would have to update the log to the latest version before retrying, right?

The core of optimistic concurrency control is we always assume we already have the latest version or latest view of the world when performing a commit. Optimistic concurrency control is optimized for low concurrent write transaction rate use-cases, so the chance of us not on the latest version is low. A lot of the time, there is only a single writer to delta table, in which case, the writer will always have the latest view of the table without needing to do explicit state refreshes. On top of this, refreshing the state here doesn't really help us get the latest view of the table, it will only get us a fairly recent view since other writes could still happen right after the refresh and before we commit the write. Therefore it's not buying us much here.

But the more important issue this leads to is correctness compromise. For some of the writes like upsert and delete, the prepared actions are only valid for a specific version of the table. For example, let's say you have a transaction to delete file A which only exists in version 1. Then subsequently another writer deletes the same file in version 2. When we try to commit the same delete action on version 3, we should abort and return a conflict error to the users instead of letting the commit go through. If we record the targeted version with the commit, the writer will know to commit on version 2 and error out with object already exists error. If we always refresh before the commit, it will fetch the latest version (2), then increment to 3 and commit without a way to perform conflict resolution.

I think leveraging execution plan is the right move here, we just need to make sure there is way to provide the targeted commit version as well.